experimental/source/source.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// package source provides standard queries and interfaces for accessing the
+// contents of source files.
+package source
+
+import (
+	"errors"
+	"io"
+	"io/fs"
+	"strings"
+
+	"github.com/bufbuild/protocompile/experimental/incremental"
+	"github.com/bufbuild/protocompile/experimental/report"
+)
+
+// Stage is the value this package uses for [report.Report].Stage.
+const Stage int = 0
+
+// Opener is a mechanism for opening files.
+//
+// Opener implementations are assumed by Protocompile to be comparable. It is
+// sufficient to always ensure that the implementation uses a pointer receiver.
+type Opener interface {
+	// Open opens a file, potentially returning an error.
+	//
+	// The result should be a string, because the syntactic analysis framework
+	// wants strings as inputs so that providing the contents of a file as a Go
+	// string minimizes copies down the line.
+	//
+	// A return value of [fs.ErrNotExist] is given special treatment by some
+	// Opener adapters, such as the [Openers] type.
+	Open(path string) (string, error)
+}
+
+// Map implements [Opener] via map lookup.
+//
+// Missing entries result in [fs.ErrNotExist].
+type Map map[string]string
```

Comment:
I think this could be error-prone -- since map is already a reference type, it could be passed around by value, and then a particular callsite using `&m` would generate a different pointer value from another (since the address-of operator would end up allocating a new heap pointer to hold a value from the stack), which would thus defeat the use of a pointer in the `Contents` query.

Maybe instead make it a struct with an unexported field and have a constructor function:
```go
type Map struct {
    sources map[string]string
}

func NewMap(m map[string]string) *Map {
    return &Map{sources: m}
}
```

Reply:
I guess. Ugh. I hate go maps lol.